File: packages/sb-uhq/src/self/commands/utilitaires/snipe.js
"use strict";

const fs   = require("fs");
const path = require("path");
const { dataPath } = require("../../func/data-path");

const WHITELIST_FILE = dataPath("msg_log_data", "snipe_whitelist.json");
const DATA_PATH      = dataPath("msg_log_data");

// ── I/O whitelist ─────────────────────────────────────────────────────────────

function getWhitelist() {
  try { return JSON.parse(fs.readFileSync(WHITELIST_FILE, "utf-8")); }
  catch { return []; }
}

function saveWhitelist(list) {
  fs.mkdirSync(path.dirname(WHITELIST_FILE), { recursive: true });
  fs.writeFileSync(WHITELIST_FILE, JSON.stringify(list, null, 2));
}

// ── Helpers lecture ───────────────────────────────────────────────────────────

function readFile(filePath) {
  try { return JSON.parse(fs.readFileSync(filePath, "utf-8")); }
  catch { return []; }
}

// ── Helpers résolution noms ───────────────────────────────────────────────────

async function resolveChannelName(client, channelId) {
  if (!channelId) return null;
  try {
    let channel = client?.channels?.cache?.get(channelId);
    if (!channel) channel = await client?.channels?.fetch(channelId).catch(() => null);
    return channel?.name ?? channel?.recipient?.tag ?? channel?.recipient?.username ?? null;
  } catch {
    return null;
  }
}

async function resolveGuildName(client, guildId) {
  if (!guildId) return null;
  try {
    let guild = client?.guilds?.cache?.get(guildId);
    if (!guild) guild = await client?.guilds?.fetch(guildId).catch(() => null);
    return guild?.name ?? null;
  } catch {
    return null;
  }
}

async function resolveUserTag(client, userId) {
  if (!userId) return null;
  try {
    let user = client?.users?.cache?.get(userId);
    if (!user) user = await client?.users?.fetch(userId).catch(() => null);
    if (!user) return null;
    if (user.discriminator && user.discriminator !== "0") {
      return `${user.username}#${user.discriminator}`;
    }
    return user.username ?? user.tag ?? null;
  } catch {
    return null;
  }
}

// ── Enrichissement des messages avec les noms de salons ──────────────────────
// Pour les recherches par guild ou user, les messages n'ont que channelId
// On résout les noms manquants en batch (un fetch par channelId unique)

async function enrichMessagesWithChannelNames(client, messages) {
  const uniqueChannelIds = [...new Set(
    messages.map(m => m.channelId).filter(Boolean)
  )];

  const channelNameMap = {};
  await Promise.all(
    uniqueChannelIds.map(async (cId) => {
      channelNameMap[cId] = await resolveChannelName(client, cId);
    })
  );

  return messages.map(m => ({
    ...m,
    channelName: m.channelName ?? channelNameMap[m.channelId] ?? null,
    // S'assurer que authorTag n'est jamais null/undefined
    authorTag: m.authorTag || "Inconnu",
    content:   m.content   ?? "",
    oldContent: m.oldContent ?? "",
    newContent: m.newContent ?? "",
  }));
}

// ── Lecture par channelId ─────────────────────────────────────────────────────

function readMessagesByChannel(channelId, type) {
  const results = [];

  for (const guildId of getWhitelist()) {
    const msgs = readFile(path.join(DATA_PATH, "SERVEURS", guildId, `${type}_messages.json`));
    results.push(...msgs.filter(m => m.channelId === channelId));
  }

  results.push(...readFile(path.join(DATA_PATH, "DMs",       channelId, `${type}_messages.json`)));
  results.push(...readFile(path.join(DATA_PATH, "GROUP_DMs", channelId, `${type}_messages.json`)));

  return results.sort((a, b) =>
    (b.deletedAt ?? b.editedAt ?? 0) - (a.deletedAt ?? a.editedAt ?? 0)
  );
}

// ── Lecture par guildId ───────────────────────────────────────────────────────

function readMessagesByGuild(guildId, type) {
  const msgs = readFile(path.join(DATA_PATH, "SERVEURS", guildId, `${type}_messages.json`));
  return msgs.sort((a, b) =>
    (b.deletedAt ?? b.editedAt ?? 0) - (a.deletedAt ?? a.editedAt ?? 0)
  );
}

// ── Lecture par userId ────────────────────────────────────────────────────────

function readMessagesByUser(userId, type) {
  const results = [];

  for (const guildId of getWhitelist()) {
    const msgs = readFile(path.join(DATA_PATH, "SERVEURS", guildId, `${type}_messages.json`));
    results.push(...msgs.filter(m => m.authorId === userId));
  }

  const dmDir = path.join(DATA_PATH, "DMs");
  if (fs.existsSync(dmDir)) {
    for (const channelId of fs.readdirSync(dmDir)) {
      const msgs = readFile(path.join(dmDir, channelId, `${type}_messages.json`));
      results.push(...msgs.filter(m => m.authorId === userId));
    }
  }

  const groupDir = path.join(DATA_PATH, "GROUP_DMs");
  if (fs.existsSync(groupDir)) {
    for (const channelId of fs.readdirSync(groupDir)) {
      const msgs = readFile(path.join(groupDir, channelId, `${type}_messages.json`));
      results.push(...msgs.filter(m => m.authorId === userId));
    }
  }

  return results.sort((a, b) =>
    (b.deletedAt ?? b.editedAt ?? 0) - (a.deletedAt ?? a.editedAt ?? 0)
  );
}

// ── Logique pure ─────────────────────────────────────────────────────────────

/**
 * @param {import("discord.js-selfbot-v13").Client} client
 * @param {{ action: string, guildId?: string, channelId?: string, userId?: string, type?: string }} payload
 */
async function execute(client, payload) {
  const { action, guildId, channelId, userId, type = "deleted" } = payload;

  if (action === "getWhitelist") {
    const whitelist = getWhitelist();
    const guilds = await Promise.all(
      whitelist.map(async (id) => ({
        id,
        name: await resolveGuildName(client, id),
      }))
    );
    return { whitelist, guilds };
  }

  if (action === "addGuild") {
    if (!guildId) throw new Error("guildId requis.");
    const list = getWhitelist();
    if (!list.includes(guildId)) { list.push(guildId); saveWhitelist(list); }
    return { whitelist: list };
  }

  if (action === "removeGuild") {
    if (!guildId) throw new Error("guildId requis.");
    const list = getWhitelist().filter(id => id !== guildId);
    saveWhitelist(list);
    return { whitelist: list };
  }

  // ── Recherche par salon ───────────────────────────────────────────────────
  if (action === "getMessages") {
    if (!channelId) throw new Error("channelId requis.");
    const rawMessages = readMessagesByChannel(channelId, type);
    const channelName = await resolveChannelName(client, channelId);
    const messages    = rawMessages.map(m => ({
      ...m,
      authorTag:  m.authorTag  || "Inconnu",
      content:    m.content    ?? "",
      oldContent: m.oldContent ?? "",
      newContent: m.newContent ?? "",
      channelName: channelName ?? null,
    }));
    return { messages, channelId, channelName, type, searchMode: "channel" };
  }

  // ── Recherche par serveur ─────────────────────────────────────────────────
  if (action === "getMessagesByGuild") {
    if (!guildId) throw new Error("guildId requis.");
    if (!getWhitelist().includes(guildId))
      throw new Error("Ce serveur n'est pas dans la whitelist.");
    const rawMessages = readMessagesByGuild(guildId, type);
    const guildName   = await resolveGuildName(client, guildId);
    const messages    = await enrichMessagesWithChannelNames(client, rawMessages);
    return { messages, guildId, guildName, type, searchMode: "guild" };
  }

  // ── Recherche par utilisateur ─────────────────────────────────────────────
  if (action === "getMessagesByUser") {
    if (!userId) throw new Error("userId requis.");
    const rawMessages = readMessagesByUser(userId, type);
    const userTag     = await resolveUserTag(client, userId);
    const messages    = await enrichMessagesWithChannelNames(client, rawMessages);
    return { messages, userId, userTag, type, searchMode: "user" };
  }

  throw new Error(`Action snipe inconnue : '${action}'`);
}

module.exports = { name: "snipe", execute };

File: packages/sb-uhq/app/commands/utilitaires/snipe.py
"""snipe — recherche dans les messages loggés (par salon / serveur / utilisateur)."""

from __future__ import annotations

import asyncio

from ...func.data_path import (data_path, is_snowflake, read_json, safe_id_segment,
                               write_json)
from ...func.discord_util import resolve_channel_name, resolve_guild_name, resolve_user_tag

WHITELIST_FILE = data_path("msg_log_data", "snipe_whitelist.json")
DATA_PATH = data_path("msg_log_data")

# Seuls types de log écrits par msglog : tout le reste deviendrait un nom de
# fichier arbitraire (`{msg_type}_messages.json`) et pourrait sortir de data/.
MSG_TYPES = ("deleted", "edited")


def _safe_msg_type(msg_type) -> str:
    if msg_type not in MSG_TYPES:
        raise ValueError("Type de messages invalide (attendu : deleted ou edited).")
    return msg_type


def get_whitelist() -> list[str]:
    # On ignore les entrées non numériques : un ID corrompu ou hérité ne doit pas
    # pouvoir servir de segment de chemin lors des recherches.
    return [g for g in read_json(WHITELIST_FILE, []) if is_snowflake(g)]


def save_whitelist(items) -> None:
    write_json(WHITELIST_FILE, items)


def _read(path) -> list:
    return read_json(path, [])


def _sort_key(m):
    return m.get("deletedAt") or m.get("editedAt") or 0


def _read_by_channel(channel_id, msg_type) -> list:
    channel_id = safe_id_segment(channel_id, "channelId")
    msg_type = _safe_msg_type(msg_type)
    results = []
    for guild_id in get_whitelist():
        msgs = _read(DATA_PATH / "SERVEURS" / guild_id / f"{msg_type}_messages.json")
        results.extend(m for m in msgs if m.get("channelId") == channel_id)
    results.extend(_read(DATA_PATH / "DMs" / channel_id / f"{msg_type}_messages.json"))
    results.extend(_read(DATA_PATH / "GROUP_DMs" / channel_id / f"{msg_type}_messages.json"))
    return sorted(results, key=_sort_key, reverse=True)


def _read_by_guild(guild_id, msg_type) -> list:
    guild_id = safe_id_segment(guild_id, "guildId")
    msg_type = _safe_msg_type(msg_type)
    msgs = _read(DATA_PATH / "SERVEURS" / guild_id / f"{msg_type}_messages.json")
    return sorted(msgs, key=_sort_key, reverse=True)


def _read_by_user(user_id, msg_type) -> list:
    user_id = safe_id_segment(user_id, "userId")
    msg_type = _safe_msg_type(msg_type)
    results = []
    for guild_id in get_whitelist():
        msgs = _read(DATA_PATH / "SERVEURS" / guild_id / f"{msg_type}_messages.json")
        results.extend(m for m in msgs if m.get("authorId") == user_id)
    for base in ("DMs", "GROUP_DMs"):
        root = DATA_PATH / base
        if root.exists():
            for channel_dir in root.iterdir():
                msgs = _read(channel_dir / f"{msg_type}_messages.json")
                results.extend(m for m in msgs if m.get("authorId") == user_id)
    return sorted(results, key=_sort_key, reverse=True)


async def _enrich(client, messages) -> list:
    unique_ids = list({m.get("channelId") for m in messages if m.get("channelId")})
    names = await asyncio.gather(*(resolve_channel_name(client, cid) for cid in unique_ids))
    name_map = dict(zip(unique_ids, names))
    enriched = []
    for m in messages:
        enriched.append({
            **m,
            "channelName": m.get("channelName") or name_map.get(m.get("channelId")),
            "authorTag": m.get("authorTag") or "Inconnu",
            "content": m.get("content") or "",
            "oldContent": m.get("oldContent") or "",
            "newContent": m.get("newContent") or "",
        })
    return enriched


async def execute(client, payload):
    action = payload.get("action")
    guild_id = payload.get("guildId")
    channel_id = payload.get("channelId")
    user_id = payload.get("userId")
    msg_type = payload.get("type", "deleted")

    if action == "getWhitelist":
        whitelist = get_whitelist()
        guilds = []
        for gid in whitelist:
            guilds.append({"id": gid, "name": await resolve_guild_name(client, gid)})
        return {"whitelist": whitelist, "guilds": guilds}

    if action == "addGuild":
        if not guild_id:
            raise ValueError("guildId requis.")
        guild_id = safe_id_segment(guild_id, "guildId")
        items = get_whitelist()
        if guild_id not in items:
            items.append(guild_id)
            save_whitelist(items)
        return {"whitelist": items}

    if action == "removeGuild":
        if not guild_id:
            raise ValueError("guildId requis.")
        items = [g for g in get_whitelist() if g != guild_id]
        save_whitelist(items)
        return {"whitelist": items}

    if action == "getMessages":
        if not channel_id:
            raise ValueError("channelId requis.")
        raw = _read_by_channel(channel_id, msg_type)
        channel_name = await resolve_channel_name(client, channel_id)
        messages = [{
            **m,
            "authorTag": m.get("authorTag") or "Inconnu",
            "content": m.get("content") or "",
            "oldContent": m.get("oldContent") or "",
            "newContent": m.get("newContent") or "",
            "channelName": channel_name,
        } for m in raw]
        return {"messages": messages, "channelId": channel_id,
                "channelName": channel_name, "type": msg_type, "searchMode": "channel"}

    if action == "getMessagesByGuild":
        if not guild_id:
            raise ValueError("guildId requis.")
        if guild_id not in get_whitelist():
            raise ValueError("Ce serveur n'est pas dans la whitelist.")
        raw = _read_by_guild(guild_id, msg_type)
        guild_name = await resolve_guild_name(client, guild_id)
        messages = await _enrich(client, raw)
        return {"messages": messages, "guildId": guild_id,
                "guildName": guild_name, "type": msg_type, "searchMode": "guild"}

    if action == "getMessagesByUser":
        if not user_id:
            raise ValueError("userId requis.")
        raw = _read_by_user(user_id, msg_type)
        tag = await resolve_user_tag(client, user_id)
        messages = await _enrich(client, raw)
        return {"messages": messages, "userId": user_id,
                "userTag": tag, "type": msg_type, "searchMode": "user"}

    raise ValueError(f"Action snipe inconnue : '{action}'")
